lib/models/package-info-cache.js

```diff
@@ -0,0 +1,810 @@
+  }
+
+  _hasErrors() {
+    return this.errors.hasErrors();
+  }
+
+  /**
+   * Add a reference to an in-repo addon PackageInfo object.
+   */
+  _addInRepoAddon(inRepoAddonPkg) {
+    if (!this.inRepoAddons) {
+      this.inRepoAddons = [];
+    }
+    this.inRepoAddons.push(inRepoAddonPkg);
+  }
+
+  /**
+   * Add a reference to an internal addon PackageInfo object.
+   * "internal" addons (note: not in-repo addons) only exist in 
+   * Projects, not other packages. Since the cache is loaded from
+   * 'loadProject', this can be done appropriately.
+   */
+  _addInternalAddon(internalAddonPkg) {
+    if (!this.internalAddons) {
+      this.internalAddons = [];
+    }
+    this.internalAddons.push(internalAddonPkg);
+  }
+
+  /**
+   * For each dependency in the given list, find the corresponding 
+   * PackageInfo object in the cache (going up the file tree if 
+   * necessary, as in the node resolution algorithm). Return a map
+   * of the dependencyName to PackageInfo object. Caller can then
+   * store it wherever they like.
+   *
+   * Note: this is not to be  called until all packages that can be have 
+   * been added to the cache. 
+   * 
+   * Note: this is for ALL dependencies, not just addons. To get just
+   * addons, filter the result by calling pkgInfo.isAddon().
+   */
+  _addDependencies(dependencies) {
+    if (!dependencies) {
+      return null;
+    }
+
+    let dependencyNames = Object.keys(dependencies);
+
+    if (dependencyNames.length === 0) {
+      return null;
+    }
+
+    let packages = Object.create(null);
+
+    let missingDependencies = [];
+    
+    dependencyNames.forEach(dependencyName => {
+      let dependencyPackage;
+
+      // much of the time the package will have dependencies in
+      // a node_modules inside it, so check there first because it's
+      // quicker since we have the reference. Only check externally
+      // if we don't find it there.
+      if (this.nodeModules) {
+        dependencyPackage = this.nodeModules._findPackage(dependencyName);
+      }
+
+      if (!dependencyPackage) {
+        dependencyPackage = this.cache._findPackage(
+          dependencyName,
+          path.dirname(this.realPath)  
+        );
+      }
+
+      if (dependencyPackage) {
+        packages[dependencyName] = dependencyPackage;
+      } else {
+        missingDependencies.push(dependencyName);
+      }
+    });
+
+    if (missingDependencies.length > 0) {
+      this._addError(ERROR_DEPENDENCIES_MISSING, missingDependencies);
+    }
+
+    return packages;
+  }
+
+  isAddon() {
+    let val =
+        Array.isArray(this.pkg.keywords) && (this.pkg.keywords.indexOf("ember-addon") >= 0);
+    return val;
+  }
+
+  // XXX Should this move to Addon.js (and change appropriately to use the cache here?)
+  getAddonConstructor() {
+    if (this.isAddon() && !this.addonConstructor) {
+      if (!this.addonMainPath) {
+        return null;
+      }
+
+      let module = require(this.addonMainPath);
```

Comment:
not for this PR, but future work. We should likely instrument the time it takes to require an addon, and set a budget. This will help ensure loading commands isn't too slow.

Reply:
Added a comment to that effect.